portal-impl/src/com/liferay/portlet/internal/PortletURLImpl.java

```diff
@@ -1191,54 +1355,86 @@ private Key _getKey() {
 		return null;
 	}
 
-	private Map<String, String[]> _mergeWithRenderParameters(
-		Map<String, String[]> portletURLParams) {
+	private LiferayMutablePortletParameters _getMutablePortletParameters(
+		String parameterName) {
 
-		String namespace = getNamespace();
+		if (_lifecycle.equals(PortletRequest.ACTION_PHASE) &&
+			!_mutableRenderParameters.isPublic(parameterName) &&
+			!parameterName.startsWith(
+				PortletQName.PRIVATE_RENDER_PARAMETER_NAMESPACE)) {
 
-		Map<String, String[]> renderParameters = RenderParametersPool.get(
-			_request, _plid, _portlet.getPortletId());
-
-		if (renderParameters == null) {
-			return portletURLParams;
+			return (LiferayMutablePortletParameters)_mutableActionParameters;
+		}
+		else if (_lifecycle.equals(PortletRequest.RESOURCE_PHASE)) {
+			return (LiferayMutablePortletParameters)_mutableResourceParameters;
 		}
 
-		Map<String, String[]> mergedRenderParams = new LinkedHashMap<>(
-			portletURLParams);
-
-		for (Map.Entry<String, String[]> entry : renderParameters.entrySet()) {
-			String name = entry.getKey();
+		return _mutableRenderParameters;
+	}
 
-			if (name.contains(namespace)) {
-				name = name.substring(namespace.length());
-			}
+	private void _initMutableRenderParameters() {
+		Set<String> publicRenderParameterNames = Collections.emptySet();
+		Map<String, String[]> mutableRenderParameterMap = null;
 
-			if (!_lifecycle.equals(PortletRequest.RESOURCE_PHASE) &&
-				(_removedParameterNames != null) &&
-				_removedParameterNames.contains(name)) {
+		if (_portletRequest == null) {
+			long plid = 0;
 
-				continue;
+			if (_layout != null) {
+				plid = _layout.getPlid();
 			}
 
-			String[] oldValues = entry.getValue();
-			String[] newValues = _params.get(name);
+			if (MimeResponse.Copy.ALL.equals(_copy) ||
+				MimeResponse.Copy.PUBLIC.equals(_copy) ||
+				_copyCurrentRenderParameters) {
+
+				Map<String, String[]> privateRenderParameterMap =
+					RenderParametersPool.get(
+						_request, plid, _portlet.getPortletId());
 
-			if (newValues == null) {
-				mergedRenderParams.put(name, oldValues);
+				if (privateRenderParameterMap != null) {
+					mutableRenderParameterMap = new HashMap<>(
+						privateRenderParameterMap);
+				}
 			}
-			else if (isBlankValue(newValues)) {
-				mergedRenderParams.remove(name);
+
+			if (mutableRenderParameterMap == null) {
+				mutableRenderParameterMap = new LinkedHashMap<>();
 			}
-			else {
-				newValues = ArrayUtil.append(newValues, oldValues);
+		}
+		else {
+			mutableRenderParameterMap = new LinkedHashMap<>();
+
+			LiferayRenderParameters renderParameters =
```

Comment:
@shuyangzhou There is a performance optimization here that I wanted to point out...

If the `_portletRequest` is *not* `null`, then that means `PortletRequestImpl` has already done the hard work of processing both public and private render parameters from the underlying `HttpServletRequest`. So instead of doing all that processing here again, I simply call `_portletRequest.getRenderParameters()`.